### DISCUSSION TITLE
Cube moves for the solver?

### DISCUSSION THREAD

**Discussion by vmiklos** [luckasRanarison/kewb#7] [Ideas]
Hi,

I guess it's known that the solver is not meant to figure out a list of moves for a random facelet state, we just want to solve the case when the facelet state is from a real-life scrambled cube. Scramble strings normally operate with the WCA notation 6 moves in 3 directions, i.e. the cube itself is never rotated. In other words we assume that the center of each face is not rotated, which means no "rotate" cube moves are done. It would be nice to support this "cube move" case, this is an interesting use-case for beginners who don't even hold the cube with the correct orientation when trying to return to the solved state, while learning (and assuming they use this solver). So far for the context. :-)

So assuming a solved state + a single F move (https://meep.cubing.net/wcanotation.html for the notation), this works:

```
$ cargo run solve --facelet UUUUUULLLURRURRURRFFFFFFFFFRRRDDDDDDLLDLLDLLDBBBBBBBBB
...
⠸ Solving
Finished in 0.19651945s
```

But this times out:

```
$ cargo run solve --facelet LUULUULUUBBBBBBBBBURRURRURRRDDRDDRDDFFFFFFFFFLLDLLDLLD
```

Possibly it does work, but the first one gives me a correct solution in ~200ms, and the second doesn't give me a solution in like 10s (didn't wait more).

Perhaps my guess is incorrect, but I assume the trouble is that the solver always tries UDLRFB moves with 3 directions, so it'll never "undo" the cube rotation (I did an "y" turn from the above page to produce the 2nd facelet string). There are a number of options to fix this problem:

1) Improve the "is solved" checker to realize that even if the center of the faces is not the expected, this is still a solved cube (if that's the cause of not finding a solution at the moment).
2) Try a few "cube turns" before the actual solving, so the cube will be always correctly oriented, as far as the current solver is concerned. This feels a bit dirty, though the problem space looks small, probably 2 turns can always recover the cube to have the wanted orientation.
3) Decide that this is out of scope for the solver and just do these cube turns in the CLI / external code that uses this crate as a library. And then just return a nicer error when this invariant is not held, instead of a timeout.

Do you have any opinion which would be the best way to handle this? Thanks.

**Comment by luckasRanarison** [comment]
Hi, 

You got an important point there, I'll also take this occasion to point the issue with solving short scrambles which I think you already noticed ;). Solvers that try to solve **every** scramble efficiently would typically solve short scrambles (less than 7 moves) with the brute force algorithm, the search space is small and it's fast enough to solve the cube in less than 1 second and it will always find solution that will never be longer than the scramble. Also the solver doesn't search in 3 directions at the moment, it's still searching sequentially (I should really implement muti-threaded search...).

Let's enter the main problem with rotations, the second facelet is the result of a `F` move + a `y` rotation right? But I think there is an issue when translating the facelet to the cubie level. Here is the resulting cubie state from the second facelet: 

```rust
    State {
        cp: [DFL, UBL, UBR, UFL, UFR, DBR, DBL, DFR],
        co: [2, 0, 0, 1, 2, 0, 0, 1],
        ep: [DF, BL, BR, UF, UL, UB, UR, FL, DR, DB, DL, FR],
        eo: [0, 1, 1, 0, 0, 0, 0, 1, 0, 0, 0, 1],
    }                                                          
```

This is an invalid (impossible) state, you'll probably notice the problem and realize what happened here if you try making `F` + `y` on a real cube (just remember that the cubie state is **relative** to the solved state). And I think the problem is the ordering of the facelet, the correct ordering is U-R-F-D-L-B and you messed the ordering when taking the rotation into account.

**Comment by luckasRanarison** [comment]
I forgot to mention, to check if a `State` is solvable or not the number of edge and corner permutations should be both odd or even. It finally makes sense to do `TryFrom<&FaceCube>` instead of just `From`, I'll update the implementation and return an Err result if that's the case because I don't plan to add support for rotations in the solver. But you've already found the solution from what I've seen :)

**Comment by vmiklos** [comment]
> Let's enter the main problem with rotations, the second facelet is the result of a `F` move + a `y` rotation right?

Correct.

> This is an invalid (impossible) state, you'll probably notice the problem and realize what happened here if you try making `F` + `y` on a real cube (just remember that the cubie state is **relative** to the solved state). And I think the problem is the ordering of the facelet, the correct ordering is U-R-F-D-L-B and you messed the ordering when taking the rotation into account.

Aha, I see. And exactly that was the question, where to handle this. So yes, if the kewb side simply detects the error and the UI is responsible to rotate the facelet string first, that's fine.

> But you've already found the solution from what I've seen :)

Right, I've [updated](https://github.com/vmiklos/vmexam/commit/396c27508b5726dbcb6a199cf94fdb393792f466) my web UI to do this rotation/normalization now.

**Comment by vmiklos** [comment]
> It finally makes sense to do TryFrom<&FaceCube> instead of just From, I'll update the implementation and return an Err result if that's the case

That sounds easy to do, just check if the middle cubie of each face is the expected "color" (e.g. U for the first one), right?

I can do that tomorrow if you want; so you can spend your time on more tricky things I would not touch for now. :-)

**Reply by luckasRanarison** [reply]
It's already done xD.

**Reply by vmiklos** [reply]
Oh, I see. Thanks!
